notifications/v1/client.go

```diff
@@ -0,0 +1,95 @@
+package v1
+
+import (
+	"context"
+	"encoding/json"
+
+	"cloud.google.com/go/pubsub"
+	"github.com/sourcegraph/log"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+// Client is a client for receiving SAMS notifications from a GCP Pub/Sub
+// subscription.
+type Client struct {
+	logger        log.Logger
+	subscription  *pubsub.Subscription
+	cancelContext context.CancelFunc
+}
+
+// NewClient creates a new Client for receiving SAMS notifications with given
+// GCP project ID and Pub/Sub subscription ID.
+func NewClient(logger log.Logger, projectID, subscriptionID string) (*Client, error) {
+	client, err := pubsub.NewClient(context.Background(), projectID)
+	if err != nil {
+		return nil, errors.Wrap(err, "create GCP Pub/Sub client")
+	}
+	return &Client{
+		logger:       logger.Scoped("notification.ClientV1"),
+		subscription: client.Subscription(subscriptionID),
+	}, nil
+}
+
+// ReceiveHandlers is a collection of receive handlers for each type of SAMS
+// notifications. If the handler of a notification is nil, the notification will
+// be acknowledged automatically without any processing.
+//
+// If a handler returns an error, the notification will be unacknowledged and
+// retried later.
+type ReceiveHandlers struct {
+	// OnUserDeleted is called when a "UserDeleted" notification is received.
+	OnUserDeleted func(data *UserDeletedData) error
```

Comment:
WDYT about just `Handler` as a name, for `notifications.Handlers.OnUserDeleted`?

Reply:
I started with `Handler`, but I prefer more concrete naming given this is after all, an argument being passed to a method (whose name is "Receive"). Also, in the future, I think it's 50/50 chance we might support Push subscription, then the naming of "Handler" might create ambiguity. WDYT?